### PR TITLE
Threads

### PR DESCRIPTION
**Pull request by jsg2021**
This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Push eslint onto a thread pool. For large projects, this helps speed it up.

### Additional Info

This is using a feature introduced in node 12... so it will fallback and disable the thread pool if it's not available.

### CODE REVIEW

**Comment by jsg2021** [comment]
This cuts my build time nearly in half. ~169s down to ~85s. :)

**Review by evilebottnawi — COMMENTED** [review]
Why do not use `jest-worker`? There all these problems have already been solved

Also I think it should be implemented on eslint side

**Comment by jsg2021** [comment]
> Why do not use `jest-worker`? All these problems have already been solved, also simple and good API

I was trying to not add dependencies. This uses just the API provided by node 12+. I also was not aware of jest-worker. I will look at it.

> Also I think it should be implemented on eslint side, but it is trade-off

We are passing one file at a time to `eslint.lintFiles()` as webpack processes dependencies... unless there are plans to make eslint parallelize that method, it would be good to support some parallelization here.

**Comment by evilebottnawi** [comment]
> I was trying to not add dependencies. This uses just the API provided by node 12+. I also was not aware of jest-worker. I will look at it.

Using `jest-worker` is recommended, here, we use this in some plugins

> We are passing one file at a time to eslint.lintFiles() as webpack processes dependencies... unless there are plans to make eslint parallelize that method, it would be good to support some parallelization here.

Yep, agree

**Comment by jsg2021** [comment]
swapped to using jest-worker.

**Review by evilebottnawi — COMMENTED** [review]
/cc @ricardogobbosouza, looks good for me

File: tsconfig.json [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "compilerOptions": {
+    "noEmit": true,
```
Comment (evilebottnawi) [inline tsconfig.json:3]: /cc @ricardogobbosouza I think here mistake

Comment (jsg2021) [reply tsconfig.json:3]: no, that was intentional. tsc was attempting to overwrite source files without that and producing an error in vscode. There is an inverse of that on the npm script generating the type definitions.

Comment (evilebottnawi) [reply tsconfig.json:3]: oh, yes, my mistake

File: src/getESLint.js [hunk 2]
```diff
@@ -1,26 +1,82 @@
...
+
+/**
+ * @param {Options} options
+ * @returns {Linter}
+ */
+export default function getESLint({ threads, ...options }) {
+  const max =
+    typeof threads !== 'number'
+      ? threads
+        ? os.cpus().length - 1
```
Comment (ksjogo) [inline src/getESLint.js:76]: What is the reasoning to use less threads than cpus?
In the old make days one would usually have done n+1 as someone might always be io blocked.

Comment (evilebottnawi) [reply src/getESLint.js:76]: One CPU for webpack, otherwise you can block webpack thread

**Pull request merged**